File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/polarity/polarity_fingerprint_v1.py
from __future__ import annotations

import hashlib

from ..canonical_json import canonical_stable_dumps

RULE4_POLARITY_ROUTING_FINGERPRINT_V1 = "rule4-polarity-routing-fingerprint-v1"


def _slot_payload(slot: dict) -> dict:
    return {
        "formula_slot_id": slot["formula_slot_id"],
        "formula_target_id": slot["formula_target_id"],
        "rule2_record_id": slot["rule2_record_id"],
        "disease_polarity": slot["disease_polarity"],
        "required_therapeutic_polarity": slot["required_therapeutic_polarity"],
        "resolution_status": slot["resolution_status"],
        "pathway": slot["pathway"],
        "potency_status": slot["potency_status"],
        "selected_cascade": slot["selected_cascade"],
        "selected_dilution": slot["selected_dilution"],
        "reason_codes": sorted(slot["reason_codes"]),
        "limitation_codes": sorted(slot["limitation_codes"]),
    }


def build_rule4_polarity_routing_fingerprint_v1_payload(
    *,
    ruleset_version: str,
    registry_version: str,
    slot_routings: list[dict],
    reason_codes: list[str],
    limitation_codes: list[str],
) -> dict:
    slots = sorted(slot_routings, key=lambda s: s["formula_slot_id"])
    return {
        "fingerprint_version": RULE4_POLARITY_ROUTING_FINGERPRINT_V1,
        "ruleset_version": ruleset_version,
        "registry_version": registry_version,
        "slot_routings": [_slot_payload(s) for s in slots],
        "reason_codes": sorted(set(reason_codes)),
        "limitation_codes": sorted(set(limitation_codes)),
    }


def rule4_polarity_routing_fingerprint_v1_hash(**kwargs) -> str:
    payload = canonical_stable_dumps(build_rule4_polarity_routing_fingerprint_v1_payload(**kwargs))
    return hashlib.sha256(payload.encode("utf-8")).hexdigest().upper()


def fingerprint_from_polarity_output(output: dict) -> str:
    return rule4_polarity_routing_fingerprint_v1_hash(
        ruleset_version=output["ruleset_version"],
        registry_version=output["registry_version"],
        slot_routings=list(output["slot_routings"]),
        reason_codes=list(output["reason_codes"]),
        limitation_codes=list(output["limitation_codes"]),
    )

File: packages/clinical-contracts/src/rule4/polarity/polarityFingerprintV1.ts
import { createHash } from 'node:crypto';

import { canonicalStableDumps } from '../canonicalJson.js';
import type { Rule4PolarityAdapterOutput, Rule4SlotPolarityRouting } from './types.js';

export const RULE4_POLARITY_ROUTING_FINGERPRINT_V1 =
  'rule4-polarity-routing-fingerprint-v1' as const;

function slotPayload(slot: Rule4SlotPolarityRouting): Record<string, unknown> {
  return {
    formula_slot_id: slot.formulaSlotId,
    formula_target_id: slot.formulaTargetId,
    rule2_record_id: slot.rule2RecordId,
    disease_polarity: slot.diseasePolarity,
    required_therapeutic_polarity: slot.requiredTherapeuticPolarity,
    resolution_status: slot.resolutionStatus,
    pathway: slot.pathway,
    potency_status: slot.potencyStatus,
    selected_cascade: slot.selectedCascade,
    selected_dilution: slot.selectedDilution,
    reason_codes: [...slot.reasonCodes].sort(),
    limitation_codes: [...slot.limitationCodes].sort(),
  };
}

export function buildRule4PolarityRoutingFingerprintV1Payload(input: {
  rulesetVersion: string;
  registryVersion: string;
  slotRoutings: readonly Rule4SlotPolarityRouting[];
  reasonCodes: readonly string[];
  limitationCodes: readonly string[];
}): Record<string, unknown> {
  const slots = [...input.slotRoutings]
    .sort((a, b) => a.formulaSlotId.localeCompare(b.formulaSlotId))
    .map(slotPayload);
  return {
    fingerprint_version: RULE4_POLARITY_ROUTING_FINGERPRINT_V1,
    ruleset_version: input.rulesetVersion,
    registry_version: input.registryVersion,
    slot_routings: slots,
    reason_codes: [...new Set(input.reasonCodes)].sort(),
    limitation_codes: [...new Set(input.limitationCodes)].sort(),
  };
}

export function rule4PolarityRoutingFingerprintV1Hash(input: {
  rulesetVersion: string;
  registryVersion: string;
  slotRoutings: readonly Rule4SlotPolarityRouting[];
  reasonCodes: readonly string[];
  limitationCodes: readonly string[];
}): string {
  const payload = canonicalStableDumps(buildRule4PolarityRoutingFingerprintV1Payload(input));
  return createHash('sha256').update(payload, 'utf8').digest('hex').toUpperCase();
}

export function fingerprintFromPolarityOutput(
  output: Pick<
    Rule4PolarityAdapterOutput,
    'rulesetVersion' | 'registryVersion' | 'slotRoutings' | 'reasonCodes' | 'limitationCodes'
  >,
): string {
  return rule4PolarityRoutingFingerprintV1Hash({
    rulesetVersion: output.rulesetVersion,
    registryVersion: output.registryVersion,
    slotRoutings: output.slotRoutings,
    reasonCodes: output.reasonCodes,
    limitationCodes: output.limitationCodes,
  });
}
